acra-censor/handlers/handlers_util.go

```diff
@@ -125,711 +171,1558 @@ func NormalizeAndRedactSQLQuery(sql string) (normalizedQuery string, redactedQue
-			}
-		}
-		if patternWhereNode, ok := patternNodes[index+patternNodeOffset].(*sqlparser.Where); ok && patternWhereNode != nil && strings.EqualFold(sqlparser.String(patternWhereNode.Expr), WhereConfigPlaceholderReplacerPart2) {
-			patternWhereDetected = true
-		}
-		if queryWhereNode, ok := queryNodes[index+queryNodeOffset].(*sqlparser.Where); ok && queryWhereNode != nil {
-			queryWhereDetected = true
-		}
-		if queryWhereDetected && patternWhereDetected {
-			return true
-		}
+	match = matchUnionLock(queryUnionNode.Lock, patternUnionNode.Lock)
+	if !match {
 		return false
 	}
-	//this is a case when pattern == query
+
 	return true
 }
-
-// handleWherePatterns try to match all WHERE conditions with supported patterns
-func handleWherePatterns(queryNodes, patternNodes []sqlparser.SQLNode) bool {
-	// collect only SelectExpr, From, Where, OrderBy ... nodes without their children
-	queryTopNodes, err := getTopNodes(queryNodes[0])
-	if err != nil {
+func handleSelectStatement(query, pattern sqlparser.Statement) bool {
+	match := false
+	querySelectNode, ok := query.(*sqlparser.Select)
+	if !ok {
 		return false
 	}
-	patternTopNodes, err := getTopNodes(patternNodes[0])
-	if err != nil {
+	patternSelectNode, ok := pattern.(*sqlparser.Select)
+	if !ok {
 		return false
 	}
-	hasStar := false
-	if selectExpr, ok := patternNodes[0].(*sqlparser.Select); ok {
-		hasStar = starFound(selectExpr.SelectExprs)
-	}
-
-	for i := 0; i < len(queryTopNodes); i++ {
-		patternNode := patternTopNodes[i]
-		queryNode := queryTopNodes[i]
-		switch patternNode.(type) {
-		case sqlparser.SelectExprs:
-			if hasStar {
-				// if select * in pattern then skip columns in query
-				continue
-			}
-		case *sqlparser.Where:
-			if _, ok := queryNode.(*sqlparser.Where); !ok {
-				return false
-			}
-			if !handleWhereNode(patternNode.(*sqlparser.Where), queryNode.(*sqlparser.Where)) {
-				return false
-			}
-			continue
-		}
 
-		if !reflect.DeepEqual(patternNode, queryNode) {
-			return false
-		}
+	// check %%SELECT%%
+	if reflect.DeepEqual(pattern, SelectPatternStatement) {
+		return true
 	}
-	return true
-}
 
-// handleLimitValuePattern evaluates LIMIT=%%VALUE%% and OFFSET=%%VALUE%% patterns
-func handleLimitValuePattern(queryNodes, patternNodes []sqlparser.SQLNode) bool {
-	querySelect, ok := queryNodes[0].(*sqlparser.Select)
-	if !ok {
+	match = matchSelectCache(querySelectNode.Cache, patternSelectNode.Cache)
+	if !match {
 		return false
 	}
-	patternSelect, ok := patternNodes[0].(*sqlparser.Select)
-	if !ok {
+	match = matchSelectComments(querySelectNode.Comments, patternSelectNode.Comments)
+	if !match {
 		return false
 	}
-	queryTopNodes, err := getTopNodes(queryNodes[0])
-	if err != nil {
+	match = matchSelectDistinct(querySelectNode.Distinct, patternSelectNode.Distinct)
+	if !match {
 		return false
 	}
-	patternTopNodes, err := getTopNodes(patternNodes[0])
-	if err != nil {
+	match = matchSelectHints(querySelectNode.Hints, patternSelectNode.Hints)
```

Comment:
as said above I prefer to add some general function that handle statements that we don't care like
```
func defaultMatch (query, pattern sqlparser.SQLNode) bool{
   return reflect.DeepEqual(query, pattern)
}
```
and then write like
```
type comparePair struct {
  pattern sqlparser.SQLNode, 
  query sqlparser.SQLNode
}
for _, pair := range []comparePair{{patternSelectNode.Distinct, querySelectNode.Distinct}, {patternSelectNode.Hints, querySelectNode.Hints}, {patternSelectNode.Comments, querySelectNode.Comments}}{
  if !defaultMatch(pair.pattern, pair.query){ return false}
}
```
and then call specific functions for correct fields as below for `SelectExpr`, `Where`, `From`